### PR TITLE
[+] Implemented user_join in usecases and UserJoinUpdatable in interface_adapters.

### PR DESCRIPTION
**Pull request by xgao28**
There are still some methods left unused, and I believe I haven't finished our UserJoinUseCase. Is there something that I need to implement next?

### CODE REVIEW

File: src/main/java/billgates/usecases/user_join/UserJoinUseCase.java
Comment (Ellen141319) [inline src/main/java/billgates/usecases/user_join/UserJoinUseCase.java:6]: I think it may not be used in your use case. Maybe you can consider to delete this one? Or you are planning to save it for future use.

File: src/main/java/billgates/usecases/user_join/UserJoinUseCase.java
Comment (Ellen141319) [inline src/main/java/billgates/usecases/user_join/UserJoinUseCase.java:21]: In my opinion, instead of hard coding something like not in the database, you should try some method calls from input.

**Review by Ellen141319 — COMMENTED** [review]
Note that there are some of my considerations in the comment.
Regarding your question, I think we should focus on the architecture and the relationship of classes in the use case more than that of the whole program. In other words, I will approve this request after you reconsidered some implementations. Overall, I think you did a good job!

File: src/main/java/billgates/usecases/user_join/UserJoinUseCase.java
Comment (xgao28) [reply src/main/java/billgates/usecases/user_join/UserJoinUseCase.java:6]: Sure!

File: src/main/java/billgates/usecases/user_join/UserJoinUseCase.java
Comment (xgao28) [reply src/main/java/billgates/usecases/user_join/UserJoinUseCase.java:21]: This could refer to one of my concerns: what do we pass in the position of "Not in database"? In other words, which of the instances could provide such key words and pass them here?

Comment (ScottCTD) [reply src/main/java/billgates/usecases/user_join/UserJoinUseCase.java:19]: If the user passed in is not in the database, we should add it to the database and login.

File: src/main/java/billgates/usecases/user_join/UserJoinUseCase.java
Comment (ScottCTD) [inline src/main/java/billgates/usecases/user_join/UserJoinUseCase.java:23]: I don't really understand why you call this function recursively.

If the user is in the database, you should pop a dialog and notify the user that the username is invalid.

**Review by ScottCTD — CHANGES_REQUESTED** [review]
The main problem is that you didn't complete the core of the use case.

Otherwise, the structure is great and style is consistent mostly.

**Comment by xgao28** [comment]
The commit above fixed some problems and I marked comments indicating such problems as resolved.

File: src/main/java/billgates/usecases/user_join/UserJoinController.java
Comment (CharlotteChenyy) [inline src/main/java/billgates/usecases/user_join/UserJoinController.java:11]: I don't know why you made this method here. We will never use it.

File: src/main/java/billgates/usecases/user_join/UserJoinController.java
Comment (CharlotteChenyy) [inline src/main/java/billgates/usecases/user_join/UserJoinController.java:15]: Same thing here, we will never use this method.

File: src/main/java/billgates/usecases/user_join/UserJoinController.java
Comment (CharlotteChenyy) [inline src/main/java/billgates/usecases/user_join/UserJoinController.java:19]: A convention error here: a method name should be starting with lower case.

Comment (CharlotteChenyy) [reply src/main/java/billgates/usecases/user_join/UserJoinController.java:19]: In addition, I don't think GUI will pass a UserJoinRequestModel to controller. Instead, it will just pass the username and password to it.

File: src/main/java/billgates/usecases/user_join/UserJoinController.java
Comment (CharlotteChenyy) [inline src/main/java/billgates/usecases/user_join/UserJoinController.java:20]: So, here we should create a UserJoinRequestModel based on the user input.

**Review by CharlotteChenyy — CHANGES_REQUESTED** [review]
I quickly reviewed your UserJoinController and found some problems

File: src/main/java/billgates/usecases/user_join/UserJoinPresenter.java
Comment (CharlotteChenyy) [inline src/main/java/billgates/usecases/user_join/UserJoinPresenter.java:5]: There should be a space before `{`

File: src/main/java/billgates/usecases/user_join/UserJoinViewModel.java
Comment (CharlotteChenyy) [inline src/main/java/billgates/usecases/user_join/UserJoinViewModel.java:8]: Should have a space before `{`

File: src/main/java/billgates/usecases/user_join/UserJoinViewModel.java
Comment (CharlotteChenyy) [inline src/main/java/billgates/usecases/user_join/UserJoinViewModel.java:18]: Should have `this` before the instance

File: src/main/java/billgates/usecases/user_join/UserJoinResponseModel.java
Comment (CharlotteChenyy) [inline src/main/java/billgates/usecases/user_join/UserJoinResponseModel.java:18]: Should have `this` before the instance

File: src/main/java/billgates/usecases/user_join/UserJoinUseCase.java
Comment (CharlotteChenyy) [inline src/main/java/billgates/usecases/user_join/UserJoinUseCase.java:17]: image
According to ray's new pull request, this should be a list of `QueryUserData`

File: src/main/java/billgates/usecases/user_join/UserJoinUseCase.java
Comment (CharlotteChenyy) [inline src/main/java/billgates/usecases/user_join/UserJoinUseCase.java:25]: What do you mean by 

> the name is invalid

?
In this else branch, if this user exists in our database, we should then check whether the input password matches with the password in our database
1. If the passwords don't match, then we should pop a dialog and notify the user that 
   1) maybe they type in an incorrect password OR 
   2) Since the user already exists in our database, we cannot register a new user with the same username

2. If the passwords match, we should then login

**Review by CharlotteChenyy — CHANGES_REQUESTED** [review]
You are on the right track right now👍👍👍
But there are still some problems.

File: src/main/java/billgates/usecases/user_join/UserJoinUseCase.java
Comment (xgao28) [reply src/main/java/billgates/usecases/user_join/UserJoinUseCase.java:17]: Should I code according to @Lei-Tin 's new commit and leave the rest? Would the auto-grading show errors?

Comment (CharlotteChenyy) [reply src/main/java/billgates/usecases/user_join/UserJoinUseCase.java:17]: You could leave your code like this right now. And we could merge your pull request first, then you could modify it after merging Ray's pull request.

Comment (CharlotteChenyy) [reply src/main/java/billgates/usecases/user_join/UserJoinUseCase.java:17]: But you have to fix the other problems that I mentioned before we merge your pull request.

File: src/main/java/billgates/usecases/user_join/UserJoinUseCase.java
Comment (CharlotteChenyy) [inline src/main/java/billgates/usecases/user_join/UserJoinUseCase.java:23]: And then we need to pop a dialog to notify the user that he/she has successfully registered.

File: src/main/java/billgates/usecases/user_join/UserJoinUseCase.java
Comment (CharlotteChenyy) [inline src/main/java/billgates/usecases/user_join/UserJoinUseCase.java:26]: This is problematic: you should not use `contains` to compare the passwords. For example, assume there is a user called `Scott` with the password `12345678`, and there is another user called `Charlotte` with the password `87654321`. What will happen if the user types in `Scott` in the `UsernameTextField`, but inputs `87654321` in the `passwordTextField`?
According to your code, you will login the user `Scott` even if the user types in the incorrect password.
Since we store the username and the corresponding password at the same index of the lists in `userData`, you could
1. Find the index of this username first
2. Then, use this index to find the corresponding password
3. Finally, compare these two passwords

**Review by CharlotteChenyy — CHANGES_REQUESTED** [review]
Everything is good, except there's still a problem with your logic of signing in

**Review by CharlotteChenyy — APPROVED** [review]
NICE👍👍👍

**Review by ScottCTD — APPROVED** [review]
Although there are some minors, it works. 

Remember to update your implementation according to the latest update from Ray.

**Pull request merged**
